hello

KinkLid opened the pull request:
pohui

leonid228322 committed:
hello

DanyaChan commented on string/str/String.h:
```diff
@@ -0,0 +1,110 @@
+public:
+	String() = default;
+
+	explicit String(const char* const& str) : size_(strlen(str) + 1), str_(new char[size_]) {
```
Давай size_ будет ровно размер строки

KinkLid replied:
ок

DanyaChan commented on string/str/String.h:
```diff
@@ -0,0 +1,110 @@
+	size_t size() const { return size_; }
+	size_t size() { return size_; }
+
+	//�� �� ������� ��� ���������
```
Комменты на английском или транслитом

KinkLid replied:
ок

DanyaChan commented on string/str/String.h:
```diff
@@ -0,0 +1,110 @@
+	}
+	//move
+	String(String&& s) : size_(s.size_), str_(s.str_) {
+		s.size_ = 0;
```
пустая строчка size_ = 0 а значит в ней -1 элемент по твоей логике

KinkLid replied:
ок

DanyaChan commented on string/str/String.h:
```diff
@@ -0,0 +1,110 @@
+	}
+
+
+
```
Нужно реализовать push_back

KinkLid replied:
готово

DanyaChan commented on string/str/str.cpp:
```diff
@@ -0,0 +1,20 @@
+#include <iostream>
+
+int main()
+{
```
Тут можно написать тесты

leonid228322 committed:
fix1

DanyaChan commented:
Пусть компилится
и пушбек добавь как в веркторе

DanyaChan commented on string/str/String.h:
```diff
@@ -0,0 +1,110 @@
+#pragma once
+#include <iostream>
```
Разбить на два файла

KinkLid replied:
сделано

KinkLid replied:
> Разбить на два файла

а вот так отвечать

leonid228322 committed:
Компилится

KinkLid committed:
Update str.cpp

Я думал так работать не будет, но работает) 
Только я не понял от куда пробел берется когда я это запускаю, типо в той ячейке которую мы "+1" делаем? но там же не должен всегда "\0" лежать как я понимаю

KinkLid committed:
fix + capacity + push_back

Фиксанул я этот пробел несчастный когда добавил капасити. и сделал функция push_back

KinkLid committed:
push_bakc fix

забыл пушить в пушбэке)

KinkLid committed:
разделение

разделение а два файла

KinkLid committed:
разделение

разделение на два файла

KinkLid committed:
fix push_back

KinkLid committed:
fix2 push_back

подумал что через while корректней будет

KinkLid committed:
sub string

добавил вывод подстроки

KinkLid committed:
sub

KinkLid committed:
Create cin_cout.h

KinkLid committed:
добавил библу cin_cout.h

DanyaChan commented on string/str/String.h:
```diff
@@ -0,0 +1,54 @@
+#pragma once
+#include <iostream>
+#include <cstring>
+using namespace std;
```
Давай без using namespace std;

KinkLid replied:
ок

DanyaChan commented on string/str/String.h:
```diff
@@ -0,0 +1,54 @@
+
+class String {
+	size_t size_ = 0;
+	size_t cap_ = 1; // capasity
```
capacity

KinkLid replied:
ок

DanyaChan commented on string/str/String.h:
```diff
@@ -0,0 +1,54 @@
+	explicit String(const char* const&);
+	String(const String&);
+	String(String&&);
+
```
добавь еще от std::string

KinkLid replied:
ok

DanyaChan commented on string/str/String.h:
```diff
@@ -0,0 +1,54 @@
+	String(String&&);
+
+	size_t size() const;
+	size_t size();
```
Что делает не конст size()?

KinkLid replied:
он на приколе

KinkLid replied:
удалил

DanyaChan commented on string/str/String.h:
```diff
@@ -0,0 +1,54 @@
+	String& operator=(const char*);
+	String& operator=(String&&);
+	String& operator+=(const String&);
+	String operator+(const String&);
```
Давай все операторы такие через friend функции, но кроме += понятно и *=

KinkLid replied:
это для того чтобы слева можно было типы перегружать для моего стринга?

KinkLid replied:
типо чтобы мой стринг можно было прировнять к инту и тд

KinkLid replied:
так ну я сделал с "+" и "*", с "=" не получилось, если такое вообще возможно

DanyaChan commented on string/str/String.h:
```diff
@@ -0,0 +1,54 @@
+	String& operator+=(const String&);
+	String operator+(const String&);
+	String operator*(int);
+	char& operator[](size_t) const;
```
пусть конст метод позвращает чар а не ссылку

DanyaChan replied:
и нужен не конст метод еще

DanyaChan commented on string/str/String.h:
```diff
@@ -0,0 +1,54 @@
+
+	void push_back(const char*);
+	String sub(size_t, size_t);
+
```
Оперпаторы сравнения нужны
Функция c_str() djpdhfoftn указатель на кост чар, типо непосредсвенно на строчку в памяти
И функция data которая уже не конст

DanyaChan replied:
И еще shrink_to_fit
всю лишнюю память освобождает, те делает capacity = size

DanyaChan commented on string/str/String.cpp:
```diff
@@ -0,0 +1,132 @@
+	String::String() = default;
+
+	String::String(const char* const& str) : size_(strlen(str)), cap_(size_ * 2), str_(new char[cap_]) {
+		memcpy(str_, str, cap_);
```
Нету смысла заранее выдеять х2 памяти

DanyaChan commented on string/str/cin_cout.h:
```diff
@@ -0,0 +1,16 @@
+
+ostream& operator<<(ostream& os, const String& s) {
+	for (size_t i = 0; i < s.size(); ++i) {
+		os << s[i];
```
Сделай строчку заканчивающуюся на \0
И просто можно будет
os << s.str_

DanyaChan commented on string/str/cin_cout.h:
```diff
@@ -0,0 +1,16 @@
+}
+
+//ne rabotaet
+istream& operator>>(istream& in, const char* c) {
```
istream& operator>>(istream& in, String& c)

такая сигнатура, читает посимвольно и делаешь пушбек одного символа

DanyaChan commented on string/str/String.cpp:
```diff
@@ -0,0 +1,132 @@
+	}
+
+	void String::push_back(const char* c) {
+		while(strlen(c) + size_ >= cap_) 
```
сделай нормально, сначала для 1 символа потому что так проще, а потом а для строчки есть +=
нуте перегрузи += для const char*

Еще strlen каждый раз заново проходит по строчке

DanyaChan commented on string/str/String.cpp:
```diff
@@ -0,0 +1,132 @@
+	char& String::operator[](size_t i) const {
+		if (i >= 0 && i <= size_)
+			return str_[i];
+		std::cout << "daleko ti polez, drug";
```
Исключение кинь а не вот это вот все

DanyaChan commented on string/str/String.cpp:
```diff
@@ -0,0 +1,132 @@
+	}
+
+	char& String::operator[](size_t i) const {
+		if (i >= 0 && i <= size_)
```
тут вроде объебка с проверкой

DanyaChan replied:
типа строка
abc
012
но size = 3

KinkLid replied:
да походу, когда капасити добавлял не поменял

KinkLid replied:
поменял

DanyaChan commented on string/str/String.cpp:
```diff
@@ -0,0 +1,132 @@
+		return ss;
+	}
+
+	String String::operator*(int a) {
```
для плюса сделал +=, а *= нет, сделай, и уже обычное * через него

KinkLid replied:
эх ладно

KinkLid replied:
сделал + починил. не правильно работало

DanyaChan commented on string/str/String.cpp:
```diff
@@ -0,0 +1,132 @@
+	}
+
+	String& String::operator+=(const String& s) {
+		cap_ = (size_ + s.size_) * 2;
```
Нету опять смысла сразу в 2 раза больше выделять, это нужно **только** в push_back

KinkLid replied:
ок

DanyaChan commented on string/str/String.cpp:
```diff
@@ -0,0 +1,132 @@
+	}
+
+	String& String::operator=(String&& s) {
+		delete[] str_;
```
copy and swap мб видел на лекция, если да, то сделай через него

KinkLid replied:
сделал

DanyaChan commented on string/str/str.cpp:
```diff
@@ -0,0 +1,20 @@
+int main()
+{
+    String s;
+    s = "AHAHAHHA";
```
Добавь всякую хуйню на сравнения, присваивание +* и тп

DanyaChan replied:
мало тестов тобишь

DanyaChan replied:
test

KinkLid replied:
на самом деле их куча у меня тут написана, на каждую хуйню что я делаю, просто я их не добавляю сюда

KinkLid committed:
push back

KinkLid committed:
push back

KinkLid committed:
std

KinkLid committed:
std

KinkLid committed:
global fix

KinkLid committed:
global fix

только теперь он мне пишет что "=" не однозначное, че делать?

KinkLid committed:
push_back fix

KinkLid committed:
cin fix

работает, но типо 255 максимум ввести, или сколько хочешь, но факт в том что я хз как сделать это динамическим
